grpc/src/main/java/io/stargate/grpc/service/Service.java

```diff
@@ -421,41 +454,60 @@ private void executePrepared(
 
       QueryParameters parameters = query.getParameters();
 
-      connection
-          .execute(
-              bindValues(handler, prepared, values), makeParameters(parameters), queryStartNanoTime)
+      CompletableFuture<Result> queryExecute =
+          connection.execute(
+              bindValues(handler, prepared.preparedQuery, values),
+              makeParameters(parameters),
+              queryStartNanoTime);
+
+      queryExecute
+          .handle(handleQuery(query, responseObserver, handler))
           .whenComplete(
-              (result, t) -> {
-                if (t != null) {
-                  handleException(t, responseObserver);
+              executeTracingQueryIfNeeded(
+                  connection, responseObserver, parameters, prepared.preparedTracing, handler));
+    } catch (Throwable t) {
+      handleException(t, responseObserver);
+    }
+  }
+
+  @NotNull
+  private BiConsumer<Response.Builder, Throwable> executeTracingQueryIfNeeded(
+      Connection connection,
+      StreamObserver<Response> responseObserver,
+      QueryParameters parameters,
+      Prepared preparedTracing,
+      PayloadHandler handler) {
+    return (responseBuilder, t) -> {
+      if (t != null) {
+        handleException(t, responseObserver);
+      } else if (!parameters.getTracing() || responseBuilder.getTracingId().isEmpty()) {
+        // tracing is not enabled or not present, fill the response observer immediately
+        Response response = responseBuilder.build();
+        responseObserver.onNext(response);
+        responseObserver.onCompleted();
+      } else {
+        try {
+          long queryStartNanoTime = System.nanoTime();
+
+          CompletableFuture<Result> queryExecute =
+              connection.execute(
+                  bindValues(handler, preparedTracing, cqlPayload(responseBuilder.getTracingId())),
```

Comment:
It was not the original intent of the payload handler logic for handling bind values and converting result sets for internal queries. It's purpose is to convert to/from the internal representation to gRPC payload values.

Perhaps it would be better to use `DataStore` for tracing/internal queries? I wonder if we could add a way to make a `DataStore` from an existing `Connection`.  If not, I don't think it's very expensive to construct a new `DataStore`.

Reply:
Regarding adding the tracing query logic to the `DataStore`, are you sure that this is the correct place?
The `DataStore` can have N implementations. We already have `PersistenceBackedDataStore` and `ValidatingDataStore`. If we decide to add a new method for fetching tracing, we would need to duplicate it. Maybe the best middle ground would be a dedicated `TraceFetcher` that takes the connection as an argument wdyt?

Reply:
The code is already using `Persistence` directly. Why use the concrete impl `PersistenceBackedDataStore`?